File: wrappers/uniffi/swift/Tests/AsdkTests/Vault.swift
import Testing
import Foundation
import Swifter
@testable import Asdk

@Suite(.serialized) class VaultTests {
    private var credentialId = "cred:12345"
        private var criteria = ["$.vct", "$.name"]
        private var credentialEntries = [
            CredentialEntry(
                credential: Credential(
                    format: VcFormat.sdJwtVc,
                    payload: "eyJ0eXAiOiJ2YytzZC1qd3QiLCJhbGciOiJFUzI1NiIsImtpZCI6ImRpZDprZXk6ekRuYWV4ZWgzVDFDemlXV1NFZVdweXVUa1hxaVQ1aWtpQ3c1aVpRUkJ2NEhYdWV4NiN6RG5hZXhlaDNUMUN6aVdXU0VlV3B5dVRrWHFpVDVpa2lDdzVpWlFSQnY0SFh1ZXg2In0.eyJfc2QiOlsiZkp1Ri1FNUMzTnhleU5UTnNMbm1DX1pnM2FNYkVwTGF1QV9aWVFnU1B3VSJdLCJ2Y3QiOiJodHRwczovL2NyZWRlbnRpYWxzLmV4YW1wbGUuY29tL2lkZW50aXR5X2NyZWRlbnRpYWwiLCJzdWIiOiJkaWQ6a2V5OnpEbmFlajlRYWRnZFpudTh1RFhaWGQ0NTQ1ZGZKQUV2bVY2bm43eGFZVXF6Y3JQdk0iLCJuYmYiOjE3Mjg4ODI2MTEsIl9zZF9hbGciOiJzaGEtMjU2IiwiaXNzIjoiZGlkOmtleTp6RG5hZXhlaDNUMUN6aVdXU0VlV3B5dVRrWHFpVDVpa2lDdzVpWlFSQnY0SFh1ZXg2IiwiaWF0IjoxNzI4ODgyNjExLCJleHAiOjE3NjA0MTg2MTEsImNuZiI6eyJqd2siOnsia3R5IjoiRUMiLCJjcnYiOiJQLTI1NiIsIngiOiJGaEFNdi1UWGcyZ1NlOGpqZkhVcWdkTzdfMjZlSG9tWVNweUxxQk05WlNZIiwieSI6IkFNelNtSXRoMHZCUTFmZjI4RlF6c1paSS1XckxZdXFxSFI4TF9HbHZrWXMifX19.usBLTsyl9fgJWPjJvbyJlpaDmfXZNRuxJCt9voME2VAAb0GhncwakNACMUdAqS9fMU5e9Y9p-KUsuOOXXVAlmg~WyI4elFmQkItS3FZSHVKcW5wVER2c1VRIiwgIm5hbWUiLCAiSm9obiJd~"
                ),
                kid: "kid",
                id: "cred:12345"
            )
        ]
        private var metadata = CredentialMetadata(
            type: "vc_type",
            format: VcFormat.sdJwtVc,
            kid: "kid",
            alg: nil,
            fields: ["$.vct", "$.name"]
        )


    @Test func getCredential() async throws {
        let vault = self.mockVault()
        let credential = try await vault.getCredential(id: self.credentialId)

        #expect(credentialEntries.first == credential)
    }
    
    @Test func getAbsentCredential() async throws {
        let vault = self.mockVault()
        let credential = try await vault.getCredential(id: "cred:56789")

        #expect(nil == credential)
    }
    
    @Test func getCredentials() async throws {
        let vault = self.mockVault()
        let credentials = try await vault.getCredentials(pagination: nil)

        #expect(self.credentialEntries == credentials)
    }
    
    @Test func findCredentials() async throws {
        let vault = self.mockVault()
        let credentials = try await vault.findCredentials(fields: self.criteria, pagination: nil)

        #expect(self.credentialEntries == credentials)
    }
    
    @Test func storeCredential() async throws {
        let vault = self.mockVault()
        let storedCredentialId = try await vault.storeCredential(credential: self.credentialEntries.first!.credential, metadata: self.metadata)

        #expect(self.credentialId == storedCredentialId)
    }
    
    @Test func deleteCredential() async throws {
        let vault = self.mockVault()
        try await vault.deleteCredential(id: self.credentialId)
    }
    
    private func mockVault() -> Asdk.Vault {
        return wrapVaultForTests(vault: MockVault(
            credentialId: self.credentialId,
            criteria: self.criteria,
            credentialEntries: self.credentialEntries,
            metadata: self.metadata
        ))
    }

}


final class MockVault : Asdk.Vault {
    
    private let credentialId: String
    private let criteria: Array<String>
    private let credentialEntries: Array<CredentialEntry>
    private let metadata: CredentialMetadata
    
    init(
        credentialId: String,
        criteria: Array<String>,
        credentialEntries: Array<CredentialEntry>,
        metadata: CredentialMetadata
    ) {
        self.credentialId = credentialId
        self.criteria = criteria
        self.credentialEntries = credentialEntries
        self.metadata = metadata
    }
    
    func storeCredential(credential: Asdk.Credential, metadata: Asdk.CredentialMetadata) async throws -> String {
        #expect(self.credentialEntries.first?.credential == credential)
        #expect(self.metadata == metadata)
        return self.credentialId
    }
    
    func deleteCredential(id: String) async throws {
        #expect(self.credentialId == id)
    }
    
    func getCredential(id: String) async throws -> Asdk.CredentialEntry? {
        if (id != self.credentialId || self.credentialEntries.isEmpty) {
            return nil
        }

        return self.credentialEntries.first
    }
    
    func getCredentials(pagination: Asdk.VaultPagination?) async throws -> [Asdk.CredentialEntry] {
        return self.credentialEntries
    }
    
    func findCredentials(fields: [String], pagination: Asdk.VaultPagination?) async throws -> [Asdk.CredentialEntry] {
        #expect(self.criteria == fields)

        return self.credentialEntries
    }
    
    
}

File: wrappers/uniffi/kotlin/src/test/kotlin/com/bci/Vault.kt
package com.bci

import com.bci.asdk.Credential
import com.bci.asdk.CredentialEntry
import com.bci.asdk.CredentialMetadata
import com.bci.asdk.VaultFetchOptions
import com.bci.asdk.VcFormat
import com.bci.asdk.setJniLibPath
import com.bci.asdk.wrapVaultForTests
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.BeforeAll
import kotlin.test.Test
import kotlin.test.assertEquals
import com.bci.asdk.Vault as ASDKVault

class Vault {
    val credentialId = "cred:12345"
    val criteria = arrayOf("$.vct", "$.name").toList()
    val credentialEntries = arrayOf(
        CredentialEntry(
            Credential(
                VcFormat.SD_JWT_VC,
                "eyJ0eXAiOiJ2YytzZC1qd3QiLCJhbGciOiJFUzI1NiIsImtpZCI6ImRpZDprZXk6ekRuYWV4ZWgzVDFDemlXV1NFZVdweXVUa1hxaVQ1aWtpQ3c1aVpRUkJ2NEhYdWV4NiN6RG5hZXhlaDNUMUN6aVdXU0VlV3B5dVRrWHFpVDVpa2lDdzVpWlFSQnY0SFh1ZXg2In0.eyJfc2QiOlsiZkp1Ri1FNUMzTnhleU5UTnNMbm1DX1pnM2FNYkVwTGF1QV9aWVFnU1B3VSJdLCJ2Y3QiOiJodHRwczovL2NyZWRlbnRpYWxzLmV4YW1wbGUuY29tL2lkZW50aXR5X2NyZWRlbnRpYWwiLCJzdWIiOiJkaWQ6a2V5OnpEbmFlajlRYWRnZFpudTh1RFhaWGQ0NTQ1ZGZKQUV2bVY2bm43eGFZVXF6Y3JQdk0iLCJuYmYiOjE3Mjg4ODI2MTEsIl9zZF9hbGciOiJzaGEtMjU2IiwiaXNzIjoiZGlkOmtleTp6RG5hZXhlaDNUMUN6aVdXU0VlV3B5dVRrWHFpVDVpa2lDdzVpWlFSQnY0SFh1ZXg2IiwiaWF0IjoxNzI4ODgyNjExLCJleHAiOjE3NjA0MTg2MTEsImNuZiI6eyJqd2siOnsia3R5IjoiRUMiLCJjcnYiOiJQLTI1NiIsIngiOiJGaEFNdi1UWGcyZ1NlOGpqZkhVcWdkTzdfMjZlSG9tWVNweUxxQk05WlNZIiwieSI6IkFNelNtSXRoMHZCUTFmZjI4RlF6c1paSS1XckxZdXFxSFI4TF9HbHZrWXMifX19.usBLTsyl9fgJWPjJvbyJlpaDmfXZNRuxJCt9voME2VAAb0GhncwakNACMUdAqS9fMU5e9Y9p-KUsuOOXXVAlmg~WyI4elFmQkItS3FZSHVKcW5wVER2c1VRIiwgIm5hbWUiLCAiSm9obiJd~"
            ),
            "kid", this.credentialId
        )
    ).toList()
    val metadata = CredentialMetadata(
        "vc_type",
        VcFormat.SD_JWT_VC,
        "kid",
        null,
        arrayOf("$.vct", "$.name").toList(),
    )

    companion object {
        @JvmStatic
        @BeforeAll
        fun setup() {
            setJniLibPath()
        }
    }

    @Test
    fun getCredential() = runTest {
        val vault = mockVault()
        val credential = vault.getCredential(credentialId)

        assertEquals(credentialEntries[0], credential)
    }

    @Test
    fun getAbsentCredential() = runTest {
        val vault = mockVault()
        val credential = vault.getCredential("cred:56789")

        assertEquals(null, credential)
    }

    @Test
    fun getCredentials() = runTest {
        val vault = mockVault()
        val credentials = vault.getCredentials(null)

        assertEquals(credentialEntries, credentials)
    }

    @Test
    fun findCredentials() = runTest {
        val vault = mockVault()
        val credentials = vault.findCredentials(criteria, null)

        assertEquals(credentialEntries, credentials)
    }

    @Test
    fun storeCredential() = runTest {
        val vault = mockVault()
        val storedCredentialId = vault.storeCredential(credentialEntries[0].credential, metadata)

        assertEquals(credentialId, storedCredentialId)
    }

    @Test
    fun deleteCredential() = runTest {
        val vault = mockVault()
        vault.deleteCredential(credentialId)
    }

    private fun mockVault(): ASDKVault {
        return wrapVaultForTests(
            MockVault(
                this.credentialId,
                this.criteria,
                this.credentialEntries,
                this.metadata
            )
        )
    }
}


class MockVault(
    private val credentialId: String,
    private val criteria: List<String>,
    private val credentialEntries: List<CredentialEntry>,
    private val metadata: CredentialMetadata
) : ASDKVault {

    override suspend fun storeCredential(
        credential: Credential,
        metadata: CredentialMetadata
    ): String {
        assertEquals(this.credentialEntries[0].credential, credential)
        assertEquals(this.metadata, metadata)
        return this.credentialId
    }

    override suspend fun deleteCredential(id: String) {
        assertEquals(this.credentialId, id)
    }

    override suspend fun getCredential(id: String): CredentialEntry? {
        if (id != this.credentialId || this.credentialEntries.isEmpty()) {
            return null
        }

        return this.credentialEntries[0]
    }

    override suspend fun getCredentials(pagination: VaultFetchOptions?): List<CredentialEntry> {
        return this.credentialEntries
    }

    override suspend fun findCredentials(
        fields: List<String>,
        pagination: VaultFetchOptions?
    ): List<CredentialEntry> {
        assertEquals(this.criteria, fields)

        return this.credentialEntries
    }

}